Use csv.DictReader when parsing wal-e backup-list

joar opened the pull request:
wal-e outputs in CSV format using the 'excel-tab' dialect:
https://github.com/wal-e/wal-e/blob/3164de68527e6ace269a2112291344b18b9ca6c5/wal_e/operator/backup.py#L63

The ISO date may be written with a space instead of 'T' as delimiter between date
and time, this causes the old parsing to fail.

joar commented:
I'm working on a fix for the failing test.

CyberDem0n commented:
I wonder, does wal-e backup-list returns tab separated values?
May be it would be easier to change `backup_strings[].split()` with `backup_strings[].split('\t')`?

joar commented:
@CyberDem0n The link in the PR points to the code where wal-e outputs the backup-list, I double-checked in a running spilo container:
```console
root@patroni-set-0005-1:/home/postgres# envdir etc/wal-e.d/env/ wal-e backup-list --detail LATEST|xxd
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "backup-list".
        STRUCTURED: time=2017-04-25T09:07:38.246711-00 pid=3978
oauth2client.transport INFO     Attempting refresh to obtain initial access_token
oauth2client.client INFO     Refreshing access_token
00000000: 6e61 6d65 096c 6173 745f 6d6f 6469 6669  name.last_modifi
00000010: 6564 0965 7870 616e 6465 645f 7369 7a65  ed.expanded_size
00000020: 5f62 7974 6573 0977 616c 5f73 6567 6d65  _bytes.wal_segme
00000030: 6e74 5f62 6163 6b75 705f 7374 6172 7409  nt_backup_start.
00000040: 7761 6c5f 7365 676d 656e 745f 6f66 6673  wal_segment_offs
00000050: 6574 5f62 6163 6b75 705f 7374 6172 7409  et_backup_start.
00000060: 7761 6c5f 7365 676d 656e 745f 6261 636b  wal_segment_back
00000070: 7570 5f73 746f 7009 7761 6c5f 7365 676d  up_stop.wal_segm
00000080: 656e 745f 6f66 6673 6574 5f62 6163 6b75  ent_offset_backu
00000090: 705f 7374 6f70 0d0a 6261 7365 5f30 3030  p_stop..base_000
000000a0: 3030 3030 4130 3030 3030 3132 4230 3030  0000A0000012B000
000000b0: 3030 3036 385f 3030 3030 3030 3430 0932  00068_00000040.2
000000c0: 3031 372d 3034 2d32 3520 3031 3a34 303a  017-04-25 01:40:
000000d0: 3031 2e31 3736 3030 302b 3030 3a30 3009  01.176000+00:00.
000000e0: 3234 3231 3934 3130 3531 3335 0930 3030  242194105135.000
000000f0: 3030 3030 4130 3030 3030 3132 4230 3030  0000A0000012B000
00000100: 3030 3036 3809 3030 3030 3030 3430 0930  00068.00000040.0
00000110: 3030 3030 3030 4130 3030 3030 3132 4230  000000A0000012B0
00000120: 3030 3030 3036 3909 3030 3630 3334 3234  0000069.00603424
00000130: 0d0a                                     ..
```
[Tab is hex `09`](https://en.wikipedia.org/wiki/Tab_key#Tab_characters)

joar commented:
@CyberDem0n I think you are correct in that it could also be fixed by using `backup_strings[].split('\t')`. I opted to use `csv.DictReader` because of the circumstances around it.

Previously:

- Split output into rows, assert # rows is `2`
- Split lines into columns, assert # columns are `7` in order to verify that parsing was done correctly.
- Convert rows of columns to dict using the first row's columns as keys and the second rows' columns as values

`csv.DictReader` does the same thing, and I suspect that it does it in a more robust way.

The benefits over just changing `.split('\t'):

- When wal-e decides to add more columns to the output, `csv.DictReader` won't make `should_use_s3[...]` fail silently.
- If wal-e would include the `\t` character in a column value it won't break our parsing of backups, additionally it will (should) be escaped on wal-e's end, and will be unescaped on our end.

joar commented:
I think this PR is ready for review now.

CyberDem0n commented on patroni/scripts/wale_restore.py:
```diff
@@ -50,18 +51,28 @@ def get_major_version(data_dir):
 
 
+WALEConfig = namedtuple('WALEConfig',
+                        'dir,threshold_mb,threshold_pct,iam_string,cmd')
```
Once you started refactoring, may be it make sense to remove `dir` and `iam_string` from `WALEConfig`?
They are not used anywhere.

Everything else seems fine.

joar replied:
Fixed in 67a9793b27a8ca9285ac19274325e5737f0a86cf

CyberDem0n commented:
:+1:

joar commented:
I pushed a fix for a bug that sneaked in during my cleanup of `WALEConfig` https://github.com/zalando/patroni/pull/436/commits/ebb042f7689c5b4a4a9c736d6e8a8385dd27a87f

CyberDem0n commented:
I've seen it.

Have you tried to execute final version on AWS?

joar commented:
My environment is Kubernetes 1.6.0 on GKE using GCS instead of S3. I don't have access to an AWS environment.

joar commented:
So, wal-e backup-list seems to be working in my environment, but I'm getting an error at another point now:

```2017-04-26 06:30:57,744 INFO: Lock owner: patroni-set-0006-0; I am patroni-set-0006-1
2017-04-26 06:30:57,751 INFO: trying to bootstrap from leader 'patroni-set-0006-0'
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "backup-list".
        STRUCTURED: time=2017-04-26T06:30:58.362701-00 pid=7246
oauth2client.transport INFO     Attempting refresh to obtain initial access_token
oauth2client.client INFO     Refreshing access_token
2017-04-26 06:30:59,064 INFO: replica has been created using wal_e
2017-04-26 06:30:59,064 INFO: bootstrapped from leader 'patroni-set-0006-0'
2017-04-26 06:30:59,068 ERROR: Exception during execution of long running task bootstrap from leader 'patroni-set-0006-0'
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/patroni/async_executor.py", line 39, in run
    wakeup = func(*args) if args else func()
  File "/usr/local/lib/python3.5/dist-packages/patroni/ha.py", line 134, in clone
    return self.state_handler.follow(node_to_follow, cluster.leader, True)
  File "/usr/local/lib/python3.5/dist-packages/patroni/postgresql.py", line 1007, in follow
    return self._do_follow(primary_conninfo, leader, recovery, timeout)
  File "/usr/local/lib/python3.5/dist-packages/patroni/postgresql.py", line 1063, in _do_follow
    self.write_recovery_conf(primary_conninfo)
  File "/usr/local/lib/python3.5/dist-packages/patroni/postgresql.py", line 899, in write_recovery_conf
    with open(self._recovery_conf, 'w') as f:
FileNotFoundError: [Errno 2] No such file or directory: '/home/postgres/pgdata/pgroot/data/recovery.conf'
```

The trace is from inside a spilo container, the spilo container has been initialized with an empty storage volume, and is expected to assume the role of replica.

I expected that patroni would start downloading basebackup or wal-e from GCS backup. Does patroni have this feature? I thought I'd read that somewhere, but https://github.com/zalando/patroni/issues/354 made me doubt.

Checking the FS of the container, I can see that:

- `/home/postgres/postgres.yml` has been created: https://gist.github.com/joar/321652e961754d8362195b3776692185
- `/home/postgres/pgdata` exists, but without `pgroot/data`:  
```
pgdata/
├── lost+found
├── pgroot
│   └── pg_log
│       ├── postgresql-0.csv
│       ├── postgresql-1.csv
│       ├── postgresql-2.csv
│       ├── postgresql-3.csv
│       ├── postgresql-4.csv
│       ├── postgresql-5.csv
│       ├── postgresql-6.csv
│       └── postgresql-7.csv
└── tmp
```

If I create `pgroot/data` manually, I get another error:
```2017-04-26 09:33:53,787 INFO success: patroni entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2017-04-26 09:33:53,847 INFO exited: patroni (exit status 1; not expected)
2017-04-26 09:33:54,850 INFO spawned: 'patroni' with pid 15819
2017-04-26 09:33:55,345 INFO: Selected new etcd server http://10.71.254.134:2379
2017-04-26 09:33:55,353 CRITICAL: system ID mismatch, node patroni-set-0006-1 belongs to a different cluster: 6412988290203516990 != 
2017-04-26 09:33:55,858 ERROR: 
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/patroni/__init__.py", line 134, in patroni_main
    patroni.run()
  File "/usr/local/lib/python3.5/dist-packages/patroni/__init__.py", line 110, in run
    logger.info(self.ha.run_cycle())
  File "/usr/local/lib/python3.5/dist-packages/patroni/ha.py", line 946, in run_cycle
    info = self._run_cycle()
  File "/usr/local/lib/python3.5/dist-packages/patroni/ha.py", line 903, in _run_cycle
    sys.exit(1)
SystemExit: 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/patroni/dcs/etcd.py", line 329, in wrapper
    retval = func(self, *args, **kwargs) is not None
  File "/usr/local/lib/python3.5/dist-packages/patroni/dcs/etcd.py", line 511, in delete_leader
    return self._client.delete(self.leader_path, prevValue=self._name)
  File "/usr/local/lib/python3.5/dist-packages/etcd/client.py", line 584, in delete
    self.key_endpoint + key, self._MDELETE, params=kwds)
  File "/usr/local/lib/python3.5/dist-packages/patroni/dcs/etcd.py", line 211, in api_execute
    return self._handle_server_response(response)
  File "/usr/local/lib/python3.5/dist-packages/etcd/client.py", line 928, in _handle_server_response
    etcd.EtcdError.handle(r)
  File "/usr/local/lib/python3.5/dist-packages/etcd/__init__.py", line 304, in handle
    raise exc(msg, payload)
etcd.EtcdCompareFailed: Compare failed : [patroni-set-0006-1 != patroni-set-0006-0]
```

CyberDem0n commented:
>    /home/postgres/postgres.yml has been created: https://gist.github.com/joar/321652e961754d8362195b3776692185

File looks good to me.

>    /home/postgres/pgdata exists 

/home/postgres/pgdata is a mountpoint

> but without pgroot/data:

That means wal-e failed to download backup from s3.
But in this case it should exit with non zero exit code...

joar commented:
> /home/postgres/pgdata is a mountpoint

I thought i mentioned that, but yes, it is.

The reason why it exits is that `self.state_handler.sysid` is `''` on https://github.com/zalando/patroni/blob/master/patroni/ha.py#L901

joar commented:
if I run `self.state_handler.create_replica(self.cluster.leader)` from within `pdb` in the context of https://github.com/zalando/patroni/blob/master/patroni/ha.py#L901, not much happens.
```
(Pdb) p self.state_handler.clone(self.cluster.leader)
2017-04-26 09:58:28,831 INFO: Removing data directory: /home/postgres/pgdata/pgroot/data
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "backup-list".
        STRUCTURED: time=2017-04-26T09:58:29.461166-00 pid=19068
oauth2client.transport INFO     Attempting refresh to obtain initial access_token
oauth2client.client INFO     Refreshing access_token
2017-04-26 09:58:30,618 INFO: replica has been created using wal_e
True
```
If i run `self.state_handler.create_replica(None)`, wal-e seems to run:
```(Pdb) p self.state_handler.create_replica(None)
2017-04-26 10:03:29,894 INFO: Removing data directory: /home/postgres/pgdata/pgroot/data
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "backup-list".
        STRUCTURED: time=2017-04-26T10:03:30.574692-00 pid=19131
oauth2client.transport INFO     Attempting refresh to obtain initial access_token
oauth2client.client INFO     Refreshing access_token
wal_e.main   INFO     MSG: starting WAL-E
        DETAIL: The subcommand is "backup-fetch".
        STRUCTURED: time=2017-04-26T10:03:32.015798-00 pid=19136
oauth2client.transport INFO     Attempting refresh to obtain initial access_token
oauth2client.client INFO     Refreshing access_token
wal_e.worker.gs.gs_worker INFO     MSG: beginning partition download
        DETAIL: The partition being downloaded is part_00000000.tar.lzo.
        HINT: The absolute GCS object is spilo/my-scope/wal/basebackups_005/base_000000010000000000000013_00000040/tar_partitions/part_00000000.tar.lzo.
        STRUCTURED: time=2017-04-26T10:03:33.115625-00 pid=19136
oauth2client.transport INFO     Attempting refresh to obtain initial access_token
oauth2client.client INFO     Refreshing access_token
2017-04-26 10:03:35,880 INFO: replica has been created using wal_e
0
```
but then, probably in the API thread as this is a running cluster with 2 members, this exception is raised:
```(Pdb) ----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 53236)
Traceback (most recent call last):
  File "/usr/lib/python3.5/socketserver.py", line 625, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/lib/python3.5/socketserver.py", line 354, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python3.5/socketserver.py", line 681, in __init__
    self.handle()
  File "/usr/lib/python3.5/http/server.py", line 422, in handle
    self.handle_one_request()
  File "/usr/lib/python3.5/http/server.py", line 410, in handle_one_request
    method()
  File "/usr/local/lib/python3.5/dist-packages/patroni/api.py", line 104, in do_GET
    self._write_status_response(status_code, response)
  File "/usr/local/lib/python3.5/dist-packages/patroni/api.py", line 62, in _write_status_response
    if patroni.postgresql.sysid:
  File "/usr/local/lib/python3.5/dist-packages/patroni/postgresql.py", line 319, in sysid
    data = self.controldata()
  File "/usr/local/lib/python3.5/dist-packages/patroni/postgresql.py", line 938, in controldata
    data = subprocess.check_output([self._pgcommand('pg_controldata'), self._data_dir])
  File "/usr/lib/python3.5/subprocess.py", line 626, in check_output
    **kwargs).stdout
  File "/usr/lib/python3.5/subprocess.py", line 693, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib/python3.5/subprocess.py", line 947, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.5/subprocess.py", line 1551, in _execute_child
    raise child_exception_type(errno_num, err_msg)
FileNotFoundError: [Errno 2] No such file or directory: 'pg_controldata'
----------------------------------------
```

joar commented:
I don't see anything in this PR that would cause those issues though, could this be a separate issue?

CyberDem0n commented on patroni/scripts/wale_restore.py:
```diff
@@ -50,33 +50,60 @@ def get_major_version(data_dir):
+                return 1
+        except Exception:
+            logger.exception("Exception when running WAL-E restore")
+            return 2
```
You've accidentally changed a logic of `run` method.
It must return 2 in case if restore from S3 doesn't make sense (when `should_use_s3_to_create_replica` returns `False`)

As a result `main` method receives `None` and `sys.exit(None)` finishes wale_restore with exit code = 0.


Please shift `return 2` on the line#106 one block to the left and it should start working.

joar replied:
Good catch!

CyberDem0n commented:
:+1:

alexeyklyukin commented:
👍

CyberDem0n commented:
Merged, thank you @joar !

joar commented:
Thank you for the help along the way!
